### PR TITLE
Add `__slots__` attributes to isolated class definitions in python.d modules.

### PR DESCRIPTION
**Pull request by Ferroin**
By default, Python stores all the instance attributes of an object in a dictionary.  THis is great for simplicity, and means you can add new attributes at runtime that were not defined by the original object.

However, in CPython (the reference implementation), dictionaries are not very memory efficient.  They consist of a series of 'slots', which consist of a pointer to the key for that slot, a pointer to the value for that key, and a handful of extra metadata. All the slots get pre-allocated, they start with 8 slots, and they expand automatically by reallocating to twice the number of slots when 2/3 of the existing slots have been used (so the expand to 16 slots when 6 have been used, 32 when 11 have been used, 64 when 22 have been used, etc).  As a result of this, objects tend to be bigger than they need to be by at least a few hundred bytes.

This is where the `__slots__` attribute comes in.  This lets you specify ahead of time an exact set of which instance attributes are usable for objects of that class, which in turn has a couple of nice benefits:

* Only slots for the attributes defined will be allocated, which means there is no overhead from unused slots.
* The number of slots never changes, which means that you never have to reallocate memory to assign an attribute, thus eliminating one source of latency found in object instantiation.
* Because the number of slots is fixed and exactly equal to the number of attribute names, there are never any hash collisions when accessing attributes, which makes attribute access more deterministic (and also marginally faster in some cases).
* Because you can't assign attributes whose name is not in `__slots__`, typos in attribute assignment for objects of that class will always result in syntax errors, instead of being potentially silent logic errors.

There are however a couple of disadvantages to using `__slots__`:

* Because of how weak references are handled in Python in general, you can't have weak references to objects whose class defines a `__slots__` attribute.  This one doesn't matter fr our usage, since we don't use weak references.
* For rather complicated reasons that even I don't entirely understand, `__slots__` doesn't work with multiple inheritance except for some very specific and not hugely useful cases.

My original plan was to adapt almost everything including the FrameworkServices and the actual module Service classes to use `__slots__` (total savings from doing this on the two systems I checked was around 16-32kB, but the systems in question don't run any of the modules likely to really benefit from this), but because of the second issue listed above, doing that does not work with the HAProxy module, so I just stuck to adding them to taxonomically isolated classes (that is, ones which are never inherited from, and either inherit from nothing, or don't inherit any instance attributes from anything).

CPython users will see a memory usage improvement of anywhere from a few dozen to a few hundred bytes per-instance of each class this touches (calculated based on CPython 3.6 source code), and a speed improvement of a few hundred CPU cycles on object instantiation and attribute access (calculated based on a combination of CPython 3.6 source code and compiled code for 64-bit x86).  Gutsy people who are running Netdata on PyPy will see no improvement, because it already does all the optimizations this triggers.

----
### TL;DR

This provides (really) small runtime memory usage and performance improvements for the python.d modules it touches, with minimal maintenance overhead.  Exact improvements are dependent on the particular Python implementation being used.

The changes for the ntpd, web_log, isc_dhcp, and hddtemp modules have been tested on actual production instances of what they monitor.  The icecast, mdstat, smartd_log, and varnish changes have been tested on trivial testing instances without any actual production data.  The nginx_plus and megacli modules have not been tested simply because I don't own the required hardware and don't' have a license respectively.  All testing was done on both CPython 2.7 and CPython 3.6.

There is a trivial one-line merge conflict between this and #4025.

### CODE REVIEW

**Comment by Ferroin** [comment]
@ktsaou I've submitted this as a series of commits with one per file being touched to try and simplify review and make it easier to remove ones that may not be wanted.  If you would prefer this to be a single commit, let me know when you're otherwise ready to merge it and I'll squash it to one commit with the information in the PR description.

**Comment by l2isbad** [comment]
Do you even test it?

```
[~]cat test.py 
class A:
    __slots__ = ('a')
    def __init__(self):
        self.a = 44

print(A().__dict__)
[~]python test.py 
{'a': 44}
```

i suggest to read carefully about [slots](https://stackoverflow.com/questions/472000/usage-of-slots)
**Slots are very tricky.** A lot of caveats.

**Comment by Ferroin** [comment]
As mentioned at the end of the description, I've tested (with either live production instances of the services being monitored, or dummy ones I set up just for testing) everything except megacli (which I don't have hardware for and therefore can't test) and nginx_plus (which I don't have a license for, and therefore also can't test).  I've not been tested with any versions of Python other than CPython 2.7 and CPython 3.6 because I have no other installed and switching to a version I don't have installed just for testing is a serious pain in the arse, but the semantics of `__slots__` have pretty much not changed at all since it was introduced in Python 2.2, so i'm not worried about other CPython versions.  PyPy is known compatible with this usage (and it's well documented that this is just a no-op there).  Not sure about Stackless, IronPython or Jython, but I seriously doubt that anybody is going to use those for Netdata.

I also specifically chose to only touch the classes I did because it avoids having to deal with pretty much any of the associated issues (they aren't inherited from, and only the ntpd ones inherit from anything (but they inherit no instance variables, so there are no issues there).

**Comment by l2isbad** [comment]
Don't you see class instance with `__slots__` has `__dict__` in my example?

This changes saves by `bytes`. I don't see any benefit @paulfantom

**Comment by Ferroin** [comment]
Yes, the instance has a `__dict__` attribute,, because the code directly accessed it.  The `__dict__` attribute will still work for read access even if you have a `__slots__` class attribute defined (for compatibility with questionable coding choices).

Correct code to check if it's there or not looks like:

```
class A:
    __slots__ = ('a',)

    def __init__(self):
        self.a = 44

print(dir(A()))
```

Which produces the following output on CPython 3.6:

    ['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', 'a']

And this on CPython 2.7:

    ['__doc__', '__init__', '__module__', '__slots__', 'a']

Because we're never accessing `__dict__` ourselves on any of these, it won't ever be there.

As far as the few dozen bytes, keep in mind that some of these classes (`Lease` for example) may have an unbounded number of instances, potentially totaling into 6 digits even on reasonable systems.  Also, as mentioned, it's nera zero maintenance overhead.

**Comment by l2isbad** [comment]
Look again to my example:
1. class instance has `__dict__`
2. class instance attribute stores in `__dict__`

Thats because `__slots__` is the **new style** classes tool.

**Comment by Ferroin** [comment]
> That's because __slots__ is the new style classes tool. (you have to inherit from object for py2 😄 )

OK, so we get no benefit on Python 2.  It also doesn't hurt anything there either, and given that Python 2.7 is EOL in ~2 years, I would argue it's not worth worrying too much about.  This brings up a second question though (why the hell is `__dict__` hidden but none of the other meta-attributes are), but that's somewhat OT.

> Ok, it makes some sense only for leases. And makes no sense for others.

`Disk` in the hddtemp plugin is similarly unbounded (there are systems with multiple hundreds of disks out there),  Same for some of the stuff in the nginx_plus module (I've seen nginx configs with well over a thousand peers for example), and everything in the smartd_log file (again, systems with multiple hundreds of disks do exist, and this one is even more likely than hddtemp because you can query SMART attributes on iSCSI and ATAoE connected storage devices).

In most cases, we're also creating new instances of these classes on _every_ collection cycle.  Object creation is _expensive_, and `__slots__` does help with that some.  A simple test case I ran last night with a dummy dhcp.leases file with just a five hundred leases and the isc_dhcp plugin running on CPython 3.6 on a Raspberry Pi (1GHz 4-core ARM CPU, 1GB or RAM) showed that the plugin ran about 4-8ms faster on average with the `__slots__` changes on all three classes than without, with having it only on the `Leases` class resulted in an average of about a quarter of a millisecond difference from the case of all classes having `__slots__`.  I didn't end up taking a memory measurement for that test, because it's rather difficult to get an accurate one when dealing with a system with so little free memory which swapps things out rather aggressively.

**Comment by l2isbad** [comment]
> OK, so we get no benefit on Python 2.

Haha, dilemma! Please inherit from `obj`, your PR is only partially working :smile: Python 2 stil a thing!

**Comment by Ferroin** [comment]
> Haha, dilemma! Please inherit from obj, your PR is only partially working 😄 Python 2 stil a thing!

That depends on the definition of working.  It doesn't break anything as is on Python 2 either.  In light of discussion in #4025 though, I'm going to wait to update this until things there are dealt with (and then just rebase on top of that when it gets merged).

**Comment by Ferroin** [comment]
> @Ferroin could you also provide baseline? "4-8ms" doesn't mean anything when we don't know for how long it actually ran.

Without the `__slots__` modification, the system in question shows about 80-120 ms run times on average (with spikes as high as 250 ms sometimes, though those correlate with similar spikes in runtime for every other collection method too).  So, translates to anywhere from 10% to ~3.33% for the average timings.  The actual performance difference on x86 systems is probably less (the Pi is ARM-based, and it's also got a couple of pathologically bad drivers that require a lot of work from the CPU), but it's still better than nothing, and the systems that are liable to benefit from this in general are lower-powered ones like the Pi.

> Unfortunately python 2 is still a thing and will be for some time probably even after EOL.
Coincidentally netdata also has 2 years, so you see that in such a long period a lot can change.

Other than RHEL derivatives, I actually rather doubt it will be around for very long.  A vast majority of the major distributions other than RHEL have already moved to Python 3 as the default (some have even gotten to the point that `python` without a version is Python 3), and pretty much all that haven't are working on doing so.  Some tight-embedded systems may continue to use it simply because they never update anything, but supporting such systems long-term is realistically unfeasible.

> @ktsaou @Ferroin This can be done on github, no need to do anything apart from clicking green "squash and merge" instead of default "Merge pull request" (more info https://help.github.com/articles/about-merge-methods-on-github/)

My main reason for offering to do it myself was that I figured that the full PR description would be preferred for the commit message, and I'm not sure if there's an easy way to get that with the GitHub interface.

**Comment by paulfantom** [comment]
> So, translates to anywhere from 10% to ~3.33% for the average timings. The actual performance difference on x86 systems is probably less

That is nice :+1: 

> Other than RHEL derivatives

So like [a fifth of the all public servers which use linux](https://w3techs.com/technologies/details/os-linux/all/all)? :stuck_out_tongue: 

> My main reason for offering to do it myself was that I figured that the full PR description would be preferred for the commit message, and I'm not sure if there's an easy way to get that with the GitHub interface.

Merge commit message is editable on github. But I understand what you want to achieve.

**Comment by Ferroin** [comment]
> So like a fifth of the all public servers which use linux?

I would care, except that Netdata is not and is likely never going to be packaged for RHEL 7, which in turn means that a decent percentage of operators won't even consider running Netdata (there's a surprising amount of 'Oh  CentOS and S, it's not packaged by RH, it must not be stable enough to be usable' mentality out there).  If we're lucky, we'll have a package in RHEL 8 (but I doubt it), which from what I hear elsewhere is supposed to move to Python 3.  CentOS might be worth caring about, but in general most people using that seem to be much better about not running half-decade old versions and generally update within a year or so of a new major version coming out, so they (and Scientific Linux users) are not likely to be an issue.

**Comment by paulfantom** [comment]
> there's a surprising amount of 'Oh CentOS and S, it's not packaged by RH, it must not be stable enough to be usable' mentality out there

> CentOS might be worth caring about

CentOS is responsible for 17% of linux public servers according to research I linked earlier. So I would care more about CentOS/Fedora than raspberry pi.

> Oh CentOS and S, it's not packaged by RH

CentOS **is** a Red Hat project. Just read legal notice https://www.centos.org/legal/

> If we're lucky, we'll have a package in RHEL 8

So what? RHEL7 has 10 years of support (so python 2 will also be supported on that system for that time). Also we won't get into official channel and I don't think we even should do it. AFAIK distributing netdata via official distro channels was a mistake, we should probably provide our own repo.

>  would care, except that Netdata is not and is likely never going to be packaged for RHEL 7

Actually packaging netdata for fedora and RHEL/CentOS 7 is second on my goal list (first is docker release). I care less about Debian family than Red Hat one.

> in general most people using that seem to be much better about not running half-decade old versions and generally update within a year or so of a new major version coming out, so they (and Scientific Linux users) are not likely to be an issue.

Proof? I don't do hearsays.

**Comment by ktsaou** [comment]
I am confused. I understand the pros are better memory management on python dictionaries. Which are the cons?

Please don't break python2 support. We want netdata to run everywhere, no matter how old a system is (when technically possible of course).

**Comment by Ferroin** [comment]
Strictly speaking, it's better memory management on Python objects by avoiding using dictionaries to store attributes.

As far as breaking Python 2, this shouldn't break anything at all.  The only versions I've found where it might 'break' are so old that we already don't work there (namely, pre 2.2 versions).

Overall though:

Pros:
* Slightly improved memory usage.  Actual savings is at most a few hundred bytes per-instance of each class, though many of the classes are not instantiated very often.  Largest savings is likely to be seen with the `isc_dhcpd` module, where it will be roughly proportionate to the number of pools and individual hosts.  Second largest savings is likely the `smartd` module when used on very large arrays of disks (on the order of hundreds or more).
* Marginally faster and more deterministic attribute access.  Using `__slots__` essentially turns a dynamic hash-table lookup for attribute access into a static one which is guaranteed to not have any hash collisions.  On most of the systems I've tested this on, it translates to a few hundred cycles, which is not much, but is still something.
* Marginally faster and more deterministic object instantiation.  Both because of faster attribute access, and the improved memory usage.  Similar to the faster attribute access, this isn't much, but it is something.

Debatable items:
* Using `__slots__` functionally enforces what attributes can exist for an object.  Attempts to assign values to attributes not defined in an object's `__slots__` will result in an error being thrown.  Some people consider this a benefit, others consider it a simple side effect with no real benefit.

Cons:
* Essentially zero effect on Python 2 (no performance benefit, and it technically _increases_ memory usage by a small amount here).
* Doesn't work properly with multiple inheritance (and thus can't really be used for the various framework services).
* Gets tricky even with single-inheritance because of how `__slots__` is inherited (in short, if you're not careful, you can end up with multiple memory locations allocated for a given attribute).  THis is largely avoided here because the only classes this is adding `__slots__` to are either isolated (no parent or child classes) or leaf classes (no child classes).
* Developers updating objects using `__slots__` have to make sure they add any new attributes they define to the object's `__slots__` attribute.

**Comment by ktsaou** [comment]
nice! thank you for the detailed report.

> Doesn't work properly with multiple inheritance (and thus can't really be used for the various framework services).

As I understand it, this may be a breaking point. But I don't really know.

@l2isbad this is your call...

**Comment by paulfantom** [comment]
> Slightly improved memory usage.
> Marginally faster and more deterministic attribute access.
> Marginally faster and more deterministic object instantiation.

vs.

> Doesn't work properly with multiple inheritance
> Gets tricky even with single-inheritance


Seems like introducing a "feature" which will heavily complicate code base with minor performance improvements. I am slightly against.

**Comment by Ferroin** [comment]
@paulfantom  Except none of the things this PR adds it to should ever need inheritance at all.  The only one which does involve inheritance isn't an issue either because of how it's being used (and that is likely to never change).  OTOH, some of these classes may have well over a thousand instances even on a reasonable system, so while it may be a small per-instance improvement, it has the potential to be big.  The DHCP lease tracking stuff is where this is likely to have the biggest impact, saving a few dozen bytes per instance there translates to a few hundred kB or more of memory savings for tracking lease counts on large networks.

**Comment by l2isbad** [comment]
`Slots` is the most non pythonic thing in `python`. I am against this changes.

Besides that this PR works only for `py2`.

> with minor performance improvements

super mega giga minor.

**Comment by ktsaou** [comment]
So, close?

**Comment by l2isbad** [comment]
i suggest to close
